progiter/progiter.py

```diff
@@ -470,29 +485,74 @@ def _iterate(self):
         # Iterating is performance sensitive, so separate both cases - where
         # 'freq' is used and checks can be fast, and where 'adjust' is used and
         # checks need more calculation. This is worth duplicating code for.
-        if self.adjust:
-            for self._iter_idx, item in gen:
-                yield item
-
-                between_idx = (self._iter_idx - self._now_idx)
-                need_display = between_idx >= self.freq
 
-                # No clue how much time has passed, the frequency may be way off.
-                # If 'freq' is too large, checking time is necessary to notice it.
-                if not need_display:
-                    between_time = default_timer() - self._now_time
-                    need_display = between_time >= self.time_thresh
-
-                if need_display:
-                    # update progress information every so often
-                    self._update_and_display_message()
-        else:
+        use_fast_path = True
+        if self.adjust:
+            use_fast_path = False
+
+            homogeneous = self.homogeneous
+            try:
+                if homogeneous == 'auto':
+                    homogeneous = False
+                    # Take a few steps in the slow path and then check to see
+                    # if we should continue or do go down the fast path.
+                    num_initial_steps = 5
+
+                    # A call to time is 50ns, we can accept the overhead if it
+                    # is only .01% of the total loop time
+                    overhead_threshold = 50e-9 * 10_000
+
+                    _check_times = []
+                    for (self._iter_idx, item), _ in zip(gen, range(num_initial_steps)):
+                        yield item
+                        self._slow_path_step_body()
+                        _check_times.append(self._between_time)
+
+                    if len(_check_times) > 1:
+                        # smallest, *middle, slowest = sorted(_check_times)
+                        slowest = max(_check_times)
+                        # We are moving fast, take the faster path
+                        if slowest < overhead_threshold:
+                            homogeneous = True
+
+                if homogeneous:
+                    raise _GottaGoFast
+
+                # Slow path where we do checks every iteration.
+                for self._iter_idx, item in gen:
+                    yield item
+                    self._slow_path_step_body()
```

Comment:
I'm certain a function call is slower than checking `self._iter_idx == num_initial_steps` in each loop. But there is no need for either drawback.
```
def slow_path_loop(gen):
    ...

for i in range(num_initial_steps):
    slow_path_loop(islice(gen, 1))
    _check_times.append(self.between_time)
```

Reply:
The point being that it is fine to add what are effectively constant overheads during the first 5 iterations, but the overheads do matter during the loop where most items are iterated.